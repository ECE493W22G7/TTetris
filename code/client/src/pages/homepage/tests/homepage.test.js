import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import MockedSocket from 'socket.io-mock';
import HomePage from 'pages/homepage/homepage';
import RoomPage from 'pages/roompage/roompage';
import HelpPage from 'pages/helppage/helppage';
import { BrowserRouter, Routes, Route, MemoryRouter } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import renderer from 'react-test-renderer';

describe("Testing Home Page", () => {
    let socket;

    beforeEach(() => {
        socket = new MockedSocket();
    })

    test("Should show up", () => {
        const tree = renderer
            .create(<MemoryRouter>
                <Routes>
                    <Route path="/" element={<HomePage socket={socket.socketClient}/>} />
                    <Route path="/room/:roomID" element={<RoomPage socket={socket.socketClient}/>} />
                    <Route path="/help" element={<HelpPage />} />
                </Routes>
            </MemoryRouter>)
            .toJSON();

        expect(tree).toMatchSnapshot();
    })

    test("Should send 'createRoom' to socket when the create button is clicked", async () => {
        const component = renderer
            .create(<MemoryRouter>
                <Routes>
                    <Route path="/" element={<HomePage socket={socket.socketClient}/>} />
                    <Route path="/room/:roomID" element={<RoomPage socket={socket.socketClient}/>} />
                    <Route path="/help" element={<HelpPage />} />
                </Routes>
            </MemoryRouter>)

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        const createRoomCallback = jest.fn();

        socket.on("create_room", createRoomCallback);

        const a = await component.root.findAllByType('button');
        a[0].props.onClick();

        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
        expect(createRoomCallback).toBeCalled();
    })

    test("Should send 'joinRoom' to socket with the inputed code when the join button is clicked", async () => {
        const component = renderer
            .create(<MemoryRouter>
                <Routes>
                    <Route path="/" element={<HomePage socket={socket.socketClient}/>} />
                    <Route path="/room/:roomID" element={<RoomPage socket={socket.socketClient}/>} />
                    <Route path="/help" element={<HelpPage />} />
                </Routes>
            </MemoryRouter>)

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        const joinRoomCallback = jest.fn();
        socket.on("join_room", joinRoomCallback);

        const a = await component.root.findAllByType('button');
        a[1].props.onClick();

        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
        expect(joinRoomCallback).toBeCalled();
    })

    test("Should send alert that the room is full on room join fail", async () => {
        const component = renderer
            .create(<MemoryRouter>
                <Routes>
                    <Route path="/" element={<HomePage socket={socket.socketClient}/>} />
                    <Route path="/room/:roomID" element={<RoomPage socket={socket.socketClient}/>} />
                    <Route path="/help" element={<HelpPage />} />
                </Routes>
            </MemoryRouter>)

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        socket.on("join_room", (roomID, done) => {
            done("full");
        });

        // mock alert, and listen for it being called
        const alertMock = jest.spyOn(window, 'alert').mockImplementation();
        const a = await component.root.findAllByType('button');
        a[1].props.onClick();

        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
        expect(alertMock).toHaveBeenCalled();
    })

    test("Should send navigate to helppage when the help button is pressed", async () => {
        const history = createMemoryHistory();
        render(<MemoryRouter history={history}>
            <Routes>
                <Route path="/" element={<HomePage socket={socket.socketClient}/>} />
                <Route path="/room/:roomID" element={<RoomPage socket={socket.socketClient}/>} />
                <Route path="/help" element={<HelpPage />} />
            </Routes>
        </MemoryRouter>);

        expect(screen.getByRole("link")).toHaveAttribute('href', '/help')
    })
})