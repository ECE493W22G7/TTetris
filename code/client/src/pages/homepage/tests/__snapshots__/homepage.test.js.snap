// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Home Page Should send 'createRoom' to socket when the create button is clicked 1`] = `
<div
  className="home-page"
>
  <div
    className="piece piece-1"
  />
  <div
    className="piece piece-2"
  />
  <div
    className="piece piece-3"
  />
  <div
    className="piece piece-4"
  />
  <div
    className="piece piece-5"
  />
  <div
    className="piece piece-6"
  />
  <div
    className="piece piece-7"
  />
  <div
    className="piece piece-8"
  />
  <div
    className="piece piece-9"
  />
  <div
    className="piece piece-10"
  />
  <p
    className="h1 text-danger font-weight-bold font-italic text-center title-margin "
  >
    Treacherous Tetris
  </p>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <button
      className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic text-center btn-space btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Create Room
    </button>
  </div>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <div
      className="input-group mb-3 text-center btn-space"
    >
      <input
        aria-describedby="basic-addon2"
        aria-label="Room ID"
        className="form-control"
        onChange={[Function]}
        placeholder="Enter RoomID"
        type="text"
      />
      <div
        className="input-group-append"
      >
        <button
          className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Join Room
        </button>
      </div>
    </div>
  </div>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <a
      className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic text-center btn-space"
      href="/help"
      onClick={[Function]}
      type="button"
    >
      Help
    </a>
  </div>
</div>
`;

exports[`Testing Home Page Should send 'createRoom' to socket when the create button is clicked 2`] = `
<div
  className="home-page"
>
  <div
    className="piece piece-1"
  />
  <div
    className="piece piece-2"
  />
  <div
    className="piece piece-3"
  />
  <div
    className="piece piece-4"
  />
  <div
    className="piece piece-5"
  />
  <div
    className="piece piece-6"
  />
  <div
    className="piece piece-7"
  />
  <div
    className="piece piece-8"
  />
  <div
    className="piece piece-9"
  />
  <div
    className="piece piece-10"
  />
  <p
    className="h1 text-danger font-weight-bold font-italic text-center title-margin "
  >
    Treacherous Tetris
  </p>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <button
      className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic text-center btn-space btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Create Room
    </button>
  </div>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <div
      className="input-group mb-3 text-center btn-space"
    >
      <input
        aria-describedby="basic-addon2"
        aria-label="Room ID"
        className="form-control"
        onChange={[Function]}
        placeholder="Enter RoomID"
        type="text"
      />
      <div
        className="input-group-append"
      >
        <button
          className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Join Room
        </button>
      </div>
    </div>
  </div>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <a
      className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic text-center btn-space"
      href="/help"
      onClick={[Function]}
      type="button"
    >
      Help
    </a>
  </div>
</div>
`;

exports[`Testing Home Page Should send 'joinRoom' to socket with the inputed code when the join button is clicked 1`] = `
<div
  className="home-page"
>
  <div
    className="piece piece-1"
  />
  <div
    className="piece piece-2"
  />
  <div
    className="piece piece-3"
  />
  <div
    className="piece piece-4"
  />
  <div
    className="piece piece-5"
  />
  <div
    className="piece piece-6"
  />
  <div
    className="piece piece-7"
  />
  <div
    className="piece piece-8"
  />
  <div
    className="piece piece-9"
  />
  <div
    className="piece piece-10"
  />
  <p
    className="h1 text-danger font-weight-bold font-italic text-center title-margin "
  >
    Treacherous Tetris
  </p>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <button
      className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic text-center btn-space btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Create Room
    </button>
  </div>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <div
      className="input-group mb-3 text-center btn-space"
    >
      <input
        aria-describedby="basic-addon2"
        aria-label="Room ID"
        className="form-control"
        onChange={[Function]}
        placeholder="Enter RoomID"
        type="text"
      />
      <div
        className="input-group-append"
      >
        <button
          className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Join Room
        </button>
      </div>
    </div>
  </div>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <a
      className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic text-center btn-space"
      href="/help"
      onClick={[Function]}
      type="button"
    >
      Help
    </a>
  </div>
</div>
`;

exports[`Testing Home Page Should send 'joinRoom' to socket with the inputed code when the join button is clicked 2`] = `
<div
  className="home-page"
>
  <div
    className="piece piece-1"
  />
  <div
    className="piece piece-2"
  />
  <div
    className="piece piece-3"
  />
  <div
    className="piece piece-4"
  />
  <div
    className="piece piece-5"
  />
  <div
    className="piece piece-6"
  />
  <div
    className="piece piece-7"
  />
  <div
    className="piece piece-8"
  />
  <div
    className="piece piece-9"
  />
  <div
    className="piece piece-10"
  />
  <p
    className="h1 text-danger font-weight-bold font-italic text-center title-margin "
  >
    Treacherous Tetris
  </p>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <button
      className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic text-center btn-space btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Create Room
    </button>
  </div>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <div
      className="input-group mb-3 text-center btn-space"
    >
      <input
        aria-describedby="basic-addon2"
        aria-label="Room ID"
        className="form-control"
        onChange={[Function]}
        placeholder="Enter RoomID"
        type="text"
      />
      <div
        className="input-group-append"
      >
        <button
          className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Join Room
        </button>
      </div>
    </div>
  </div>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <a
      className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic text-center btn-space"
      href="/help"
      onClick={[Function]}
      type="button"
    >
      Help
    </a>
  </div>
</div>
`;

exports[`Testing Home Page Should send alert that the room is full on room join fail 1`] = `
<div
  className="home-page"
>
  <div
    className="piece piece-1"
  />
  <div
    className="piece piece-2"
  />
  <div
    className="piece piece-3"
  />
  <div
    className="piece piece-4"
  />
  <div
    className="piece piece-5"
  />
  <div
    className="piece piece-6"
  />
  <div
    className="piece piece-7"
  />
  <div
    className="piece piece-8"
  />
  <div
    className="piece piece-9"
  />
  <div
    className="piece piece-10"
  />
  <p
    className="h1 text-danger font-weight-bold font-italic text-center title-margin "
  >
    Treacherous Tetris
  </p>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <button
      className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic text-center btn-space btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Create Room
    </button>
  </div>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <div
      className="input-group mb-3 text-center btn-space"
    >
      <input
        aria-describedby="basic-addon2"
        aria-label="Room ID"
        className="form-control"
        onChange={[Function]}
        placeholder="Enter RoomID"
        type="text"
      />
      <div
        className="input-group-append"
      >
        <button
          className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Join Room
        </button>
      </div>
    </div>
  </div>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <a
      className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic text-center btn-space"
      href="/help"
      onClick={[Function]}
      type="button"
    >
      Help
    </a>
  </div>
</div>
`;

exports[`Testing Home Page Should send alert that the room is full on room join fail 2`] = `
<div
  className="home-page"
>
  <div
    className="piece piece-1"
  />
  <div
    className="piece piece-2"
  />
  <div
    className="piece piece-3"
  />
  <div
    className="piece piece-4"
  />
  <div
    className="piece piece-5"
  />
  <div
    className="piece piece-6"
  />
  <div
    className="piece piece-7"
  />
  <div
    className="piece piece-8"
  />
  <div
    className="piece piece-9"
  />
  <div
    className="piece piece-10"
  />
  <p
    className="h1 text-danger font-weight-bold font-italic text-center title-margin "
  >
    Treacherous Tetris
  </p>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <button
      className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic text-center btn-space btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Create Room
    </button>
  </div>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <div
      className="input-group mb-3 text-center btn-space"
    >
      <input
        aria-describedby="basic-addon2"
        aria-label="Room ID"
        className="form-control"
        onChange={[Function]}
        placeholder="Enter RoomID"
        type="text"
      />
      <div
        className="input-group-append"
      >
        <button
          className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Join Room
        </button>
      </div>
    </div>
  </div>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <a
      className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic text-center btn-space"
      href="/help"
      onClick={[Function]}
      type="button"
    >
      Help
    </a>
  </div>
</div>
`;

exports[`Testing Home Page Should show up 1`] = `
<div
  className="home-page"
>
  <div
    className="piece piece-1"
  />
  <div
    className="piece piece-2"
  />
  <div
    className="piece piece-3"
  />
  <div
    className="piece piece-4"
  />
  <div
    className="piece piece-5"
  />
  <div
    className="piece piece-6"
  />
  <div
    className="piece piece-7"
  />
  <div
    className="piece piece-8"
  />
  <div
    className="piece piece-9"
  />
  <div
    className="piece piece-10"
  />
  <p
    className="h1 text-danger font-weight-bold font-italic text-center title-margin "
  >
    Treacherous Tetris
  </p>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <button
      className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic text-center btn-space btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Create Room
    </button>
  </div>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <div
      className="input-group mb-3 text-center btn-space"
    >
      <input
        aria-describedby="basic-addon2"
        aria-label="Room ID"
        className="form-control"
        onChange={[Function]}
        placeholder="Enter RoomID"
        type="text"
      />
      <div
        className="input-group-append"
      >
        <button
          className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Join Room
        </button>
      </div>
    </div>
  </div>
  <div
    className="d-flex justify-content-center align-items-start"
  >
    <a
      className="btn btn-secondary btn-lg btn-block text-dark font-weight-bold font-italic text-center btn-space"
      href="/help"
      onClick={[Function]}
      type="button"
    >
      Help
    </a>
  </div>
</div>
`;
